Fix non-compiling tests

### Context
Due to the changes introduced by the ongoing refactoring, EvoSuite's own unit tests no longer compile. The main reason for this is the changed class signature of `Chromosome`. It now uses a recursive type parameter (`Chromosome<T extends Chromosome<T>>`). This means that every unit test using a `Chromosome` must no properly instantiate this recursive type parameter. (In particular, raw types and any abstract classes such as `Chromosome` itself do no longer work.) Otherwise, you get a compile error such as 
```
Type parameter 'T' is not within its bound; should extend 'org.evosuite.ga.Chromosome<T>'
```

### Steps to Reproduce
1. Pull the latest changes from the `selfType` branch.
2. Compile the tests

### Current Result
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project evosuite-master: Compilation failure: Compilation failure: 
[ERROR] /home/schweikl/.../evosuite/master/src/test/java/org/evosuite/ga/problems/singleobjective/ThreeHump.java:[39,34] type argument T is not within bounds of type-variable T
[ERROR] /home/schweikl/.../evosuite/master/src/test/java/org/evosuite/ga/problems/singleobjective/ThreeHump.java:[39,87] type argument T is not within bounds of type-variable T
```
(among others, which are omitted for brevity)
 
 
### Expected result
The tests should compile.



### Additional info
Most of the time, fixing the tests is trivial. Consider the class `ThreeHump` in package `org.evosuite.ga.problems.singleobjective`. The relevant parts are marked with `XXX`:
```
public class ThreeHump<T extends NSGAChromosome> implements Problem  // XXX: raw use of parameterized class "Problem"
{
    private List<FitnessFunction<T>> fitnessFunctions = new ArrayList<FitnessFunction<T>>();  // XXX: Type parameter "T" not within its bound

    public ThreeHump()
    {
        super();

        /**
         * Fitness function
         */
        class fFitnessFunction extends FitnessFunction { // XXX: raw use of parameterized class "FitnessFunction"
            @Override
            public double getFitness(Chromosome c) { // XXX: raw use of parameterized class "Chromosome"
                NSGAChromosome individual = (NSGAChromosome)c;  // XXX: unchecked down-cast
                // ...
            }
        }
    }

    @Override
    public List getFitnessFunctions() { // XXX: raw use of parameterized class "List"
        return this.fitnessFunctions;
    }
```
Here, the class does not compile because type parameter `T` is not within its bounds. Moreover, the code uses raw types, which is a code smell and bad practice. These issues can be fixed as follows:
```
public class ThreeHump implements Problem<NSGAChromosome> // Specify proper type bound, use NSGAChromosome instead of T
{
    private List<FitnessFunction<NSGAChromosome>> fitnessFunctions = new ArrayList<>(); // Use NSGAChromosome instead of T

    public ThreeHump()
    {
        super();

        /**
         * Fitness function
         */
        class fFitnessFunction extends FitnessFunction<NSGAChromosome> { // Proper parameterization of FitnessFunction
            @Override
            public double getFitness(NSGAChromosome c) {  // Proper parameterization to avoid unchecked down-cast
                // No unchecked down-cast required anymore!
                // ...
            }
        }
    }

    @Override
    public List<FitnessFunction<NSGAChromosome>> getFitnessFunctions() { // Proper parameterization of List
        return this.fitnessFunctions;
    }
}
```
Since almost all tests are very similar in their structure, this strategy (as presented above) can be applied to almost all tests.

--- a/master/src/test/java/org/evosuite/SystemTestBase.java
+++ b/master/src/test/java/org/evosuite/SystemTestBase.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright (C) 2010-2018 Gordon Fraser, Andrea Arcuri and EvoSuite
  * contributors
  *

--- a/master/src/test/java/org/evosuite/ga/comparators/TestDominanceComparator.java
+++ b/master/src/test/java/org/evosuite/ga/comparators/TestDominanceComparator.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright (C) 2010-2018 Gordon Fraser, Andrea Arcuri and EvoSuite
  * contributors
  *
@@ -40,7 +40,7 @@ public class TestDominanceComparator
     @Test
     public void testDominanceComparatorOneFitness()
     {
-        Problem p = new Booths<NSGAChromosome>();
+        Problem<NSGAChromosome> p = new Booths();
         List<FitnessFunction<NSGAChromosome>> fitnessFunctions = p.getFitnessFunctions();
         FitnessFunction<NSGAChromosome> ff = fitnessFunctions.get(0);
 
@@ -55,8 +55,7 @@ public class TestDominanceComparator
         population.add(c1);
         population.add(c2);
 
-        DominanceComparator dc = new DominanceComparator();
-        Collections.sort(population, dc);
+        population.sort(new DominanceComparator<>());
 
         Assert.assertEquals(population.get(0).getFitness(ff), 0.3, 0.0);
         Assert.assertEquals(population.get(1).getFitness(ff), 0.7, 0.0);
@@ -65,7 +64,7 @@ public class TestDominanceComparator
     @Test
     public void testDominanceComparatorSeveralFitnessesNoDomination()
     {
-        Problem p = new FON();
+        Problem<NSGAChromosome> p = new FON();
         List<FitnessFunction<NSGAChromosome>> fitnessFunctions = p.getFitnessFunctions();
         FitnessFunction<NSGAChromosome> ff_1 = fitnessFunctions.get(0);
         FitnessFunction<NSGAChromosome> ff_2 = fitnessFunctions.get(1);
@@ -83,8 +82,7 @@ public class TestDominanceComparator
         population.add(c1);
         population.add(c2);
 
-        DominanceComparator dc = new DominanceComparator();
-        Collections.sort(population, dc);
+        population.sort(new DominanceComparator<>());
 
         Assert.assertEquals(population.get(0).getFitness(ff_1), 0.7, 0.0);
         Assert.assertEquals(population.get(0).getFitness(ff_2), 0.2, 0.0);
@@ -95,7 +93,7 @@ public class TestDominanceComparator
     @Test
     public void testDominanceComparatorSeveralFitnessesDomination()
     {
-        Problem p = new FON();
+        Problem<NSGAChromosome> p = new FON();
         List<FitnessFunction<NSGAChromosome>> fitnessFunctions = p.getFitnessFunctions();
         FitnessFunction<NSGAChromosome> ff_1 = fitnessFunctions.get(0);
         FitnessFunction<NSGAChromosome> ff_2 = fitnessFunctions.get(1);
@@ -113,8 +111,7 @@ public class TestDominanceComparator
         population.add(c1);
         population.add(c2);
 
-        DominanceComparator dc = new DominanceComparator();
-        Collections.sort(population, dc);
+        population.sort(new DominanceComparator<>());
 
         Assert.assertEquals(population.get(0).getFitness(ff_1), 0.3, 0.0);
         Assert.assertEquals(population.get(0).getFitness(ff_2), 0.5, 0.0);

--- a/master/src/test/java/org/evosuite/ga/comparators/TestSortByFitness.java
+++ b/master/src/test/java/org/evosuite/ga/comparators/TestSortByFitness.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright (C) 2010-2018 Gordon Fraser, Andrea Arcuri and EvoSuite
  * contributors
  *
@@ -19,16 +19,15 @@
  */
 package org.evosuite.ga.comparators;
 
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.List;
-
 import org.evosuite.ga.FitnessFunction;
 import org.evosuite.ga.NSGAChromosome;
 import org.evosuite.ga.problems.Problem;
 import org.evosuite.ga.problems.singleobjective.Booths;
 import org.junit.Assert;
 import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.List;
 
 /**
  * 
@@ -39,7 +38,7 @@ public class TestSortByFitness
     @Test
     public void testSortByFitnessC1win()
     {
-        Problem p = new Booths<NSGAChromosome>();
+        Problem<NSGAChromosome> p = new Booths();
         List<FitnessFunction<NSGAChromosome>> fitnessFunctions = p.getFitnessFunctions();
         FitnessFunction<NSGAChromosome> ff = fitnessFunctions.get(0);
 
@@ -50,24 +49,23 @@ public class TestSortByFitness
         c1.setFitness(ff, 0.7);
         c2.setFitness(ff, 0.3);
 
-        List<NSGAChromosome> population = new ArrayList<NSGAChromosome>();
+        List<NSGAChromosome> population = new ArrayList<>();
         population.add(c1);
         population.add(c2);
 
-        SortByFitness sf = new SortByFitness(ff, true);
-        Collections.sort(population, sf);
+        population.sort(new SortByFitness<>(ff, true));
 
-        Assert.assertTrue(population.get(0) == c1);
+        Assert.assertSame(population.get(0), c1);
         Assert.assertEquals(population.get(0).getFitness(ff), 0.7, 0.0);
 
-        Assert.assertTrue(population.get(1) == c2);
+        Assert.assertSame(population.get(1), c2);
         Assert.assertEquals(population.get(1).getFitness(ff), 0.3, 0.0);
     }
 
     @Test
     public void testSortByFitnessC2win()
     {
-        Problem p = new Booths<NSGAChromosome>();
+        Problem<NSGAChromosome> p = new Booths();
         List<FitnessFunction<NSGAChromosome>> fitnessFunctions = p.getFitnessFunctions();
         FitnessFunction<NSGAChromosome> ff = fitnessFunctions.get(0);
 
@@ -78,24 +76,23 @@ public class TestSortByFitness
         c1.setFitness(ff, 0.3);
         c2.setFitness(ff, 0.7);
 
-        List<NSGAChromosome> population = new ArrayList<NSGAChromosome>();
+        List<NSGAChromosome> population = new ArrayList<>();
         population.add(c1);
         population.add(c2);
 
-        SortByFitness sf = new SortByFitness(ff, true);
-        Collections.sort(population, sf);
+        population.sort(new SortByFitness<>(ff, true));
 
-        Assert.assertTrue(population.get(0) == c2);
+        Assert.assertSame(population.get(0), c2);
         Assert.assertEquals(population.get(0).getFitness(ff), 0.7, 0.0);
 
         Assert.assertEquals(population.get(1).getFitness(ff), 0.3, 0.0);
-        Assert.assertTrue(population.get(1) == c1);
+        Assert.assertSame(population.get(1), c1);
     }
 
     @Test
     public void testSortByFitnessEqual()
     {
-        Problem p = new Booths<NSGAChromosome>();
+        Problem<NSGAChromosome> p = new Booths();
         List<FitnessFunction<NSGAChromosome>> fitnessFunctions = p.getFitnessFunctions();
         FitnessFunction<NSGAChromosome> ff = fitnessFunctions.get(0);
 
@@ -106,17 +103,16 @@ public class TestSortByFitness
         c1.setFitness(ff, 0.5);
         c2.setFitness(ff, 0.5);
 
-        List<NSGAChromosome> population = new ArrayList<NSGAChromosome>();
+        List<NSGAChromosome> population = new ArrayList<>();
         population.add(c1);
         population.add(c2);
 
-        SortByFitness sf = new SortByFitness(ff, true);
-        Collections.sort(population, sf);
+        population.sort(new SortByFitness<>(ff, true));
 
-        Assert.assertTrue(population.get(0) == c1);
+        Assert.assertSame(population.get(0), c1);
         Assert.assertEquals(population.get(0).getFitness(ff), 0.5, 0.0);
 
         Assert.assertEquals(population.get(1).getFitness(ff), 0.5, 0.0);
-        Assert.assertTrue(population.get(1) == c2);
+        Assert.assertSame(population.get(1), c2);
     }
 }

--- a/master/src/test/java/org/evosuite/ga/metaheuristics/NSGAIISystemTest.java
+++ b/master/src/test/java/org/evosuite/ga/metaheuristics/NSGAIISystemTest.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright (C) 2010-2018 Gordon Fraser, Andrea Arcuri and EvoSuite
  * contributors
  *
@@ -54,16 +54,16 @@ public class NSGAIISystemTest extends SystemTestBase
 	@BeforeClass
 	public static void setUp() {
 		Properties.CROSSOVER_RATE = 0.9;
-		Properties.RANDOM_SEED = 1l;
+		Properties.RANDOM_SEED = 1L;
 	}
 
 	@Test
 	public void testUnionEmptyPopulation()
 	{
-	    NSGAII<NSGAChromosome> ga = new NSGAII<NSGAChromosome>(null);
+	    NSGAII<NSGAChromosome> ga = new NSGAII<>(null);
 
-	    List<NSGAChromosome> pop = new ArrayList<NSGAChromosome>();
-	    List<NSGAChromosome> off = new ArrayList<NSGAChromosome>();
+	    List<NSGAChromosome> pop = new ArrayList<>();
+	    List<NSGAChromosome> off = new ArrayList<>();
 	    List<NSGAChromosome> union = ga.union(pop, off);
 
 	    Assert.assertTrue(union.isEmpty());
@@ -72,17 +72,17 @@ public class NSGAIISystemTest extends SystemTestBase
 	@Test
     public void testUnion()
     {
-        NSGAII<NSGAChromosome> ga = new NSGAII<NSGAChromosome>(null);
+        NSGAII<NSGAChromosome> ga = new NSGAII<>(null);
 
         NSGAChromosome c1 = new NSGAChromosome();
         NSGAChromosome c2 = new NSGAChromosome();
         NSGAChromosome c3 = new NSGAChromosome();
 
-        List<NSGAChromosome> pop = new ArrayList<NSGAChromosome>();
+        List<NSGAChromosome> pop = new ArrayList<>();
         pop.add(c1);
         pop.add(c2);
 
-        List<NSGAChromosome> off = new ArrayList<NSGAChromosome>();
+        List<NSGAChromosome> off = new ArrayList<>();
         off.add(c3);
 
         List<NSGAChromosome> union = ga.union(pop, off);
@@ -92,9 +92,9 @@ public class NSGAIISystemTest extends SystemTestBase
 	@Test
 	public void testFastNonDominatedSort()
 	{
-		NSGAII<NSGAChromosome> ga = new NSGAII<NSGAChromosome>(null);
+		NSGAII<NSGAChromosome> ga = new NSGAII<>(null);
 
-		Problem p = new Booths<NSGAChromosome>();
+		Problem<NSGAChromosome> p = new Booths();
         List<FitnessFunction<NSGAChromosome>> fitnessFunctions = p.getFitnessFunctions();
         ga.addFitnessFunctions(fitnessFunctions);
 
@@ -121,7 +121,7 @@ public class NSGAIISystemTest extends SystemTestBase
 		c9.setFitness(fitnessFunctions.get(0), 0.6);
 		c10.setFitness(fitnessFunctions.get(0), 0.0);
 
-		List<NSGAChromosome> population = new ArrayList<NSGAChromosome>();
+		List<NSGAChromosome> population = new ArrayList<>();
 		population.add(c1);
 		population.add(c2);
 		population.add(c3);
@@ -133,38 +133,38 @@ public class NSGAIISystemTest extends SystemTestBase
 		population.add(c9);
 		population.add(c10);
 
-		ga.getRankingFunction().computeRankingAssignment(population, new LinkedHashSet<FitnessFunction<NSGAChromosome>>(fitnessFunctions));
+		ga.getRankingFunction().computeRankingAssignment(population, new LinkedHashSet<>(fitnessFunctions));
 
 		// Total number of Fronts
 		Assert.assertEquals(ga.getRankingFunction().getNumberOfSubfronts(), 5);
 
 		// Front 0
-		Assert.assertTrue(ga.getRankingFunction().getSubfront(0).get(0).getFitness() == 0.0);
+		Assert.assertEquals(0.0, ga.getRankingFunction().getSubfront(0).get(0).getFitness(), 0.0);
 		//Assert.assertTrue(ga.getRankingFunction().getSubfront(0).get(1).getFitness() == 0.0);
 
 		// Front 1
-		Assert.assertTrue(ga.getRankingFunction().getSubfront(1).get(0).getFitness() == 0.2);
-		Assert.assertTrue(ga.getRankingFunction().getSubfront(1).get(1).getFitness() == 0.2);
+		Assert.assertEquals(0.2, ga.getRankingFunction().getSubfront(1).get(0).getFitness(), 0.0);
+		Assert.assertEquals(0.2, ga.getRankingFunction().getSubfront(1).get(1).getFitness(), 0.0);
 
 		// Front 2
-		Assert.assertTrue(ga.getRankingFunction().getSubfront(2).get(0).getFitness() == 0.4);
-		Assert.assertTrue(ga.getRankingFunction().getSubfront(2).get(1).getFitness() == 0.4);
+		Assert.assertEquals(0.4, ga.getRankingFunction().getSubfront(2).get(0).getFitness(), 0.0);
+		Assert.assertEquals(0.4, ga.getRankingFunction().getSubfront(2).get(1).getFitness(), 0.0);
 
 		// Front 3
-		Assert.assertTrue(ga.getRankingFunction().getSubfront(3).get(0).getFitness() == 0.6);
-		Assert.assertTrue(ga.getRankingFunction().getSubfront(3).get(1).getFitness() == 0.6);
+		Assert.assertEquals(0.6, ga.getRankingFunction().getSubfront(3).get(0).getFitness(), 0.0);
+		Assert.assertEquals(0.6, ga.getRankingFunction().getSubfront(3).get(1).getFitness(), 0.0);
 
 		// Front 4
-		Assert.assertTrue(ga.getRankingFunction().getSubfront(4).get(0).getFitness() == 0.8);
-		Assert.assertTrue(ga.getRankingFunction().getSubfront(4).get(1).getFitness() == 0.8);
+		Assert.assertEquals(0.8, ga.getRankingFunction().getSubfront(4).get(0).getFitness(), 0.0);
+		Assert.assertEquals(0.8, ga.getRankingFunction().getSubfront(4).get(1).getFitness(), 0.0);
 	}
 
 	@Test
 	public void testCrowingDistanceAssignment_OneVariable()
 	{
-		NSGAII<NSGAChromosome> ga = new NSGAII<NSGAChromosome>(null);
+		NSGAII<NSGAChromosome> ga = new NSGAII<>(null);
 
-		Problem p = new Booths();
+		Problem<NSGAChromosome> p = new Booths();
 		List<FitnessFunction<NSGAChromosome>> fitnessFunctions = p.getFitnessFunctions();
 		ga.addFitnessFunctions(fitnessFunctions);
 
@@ -191,7 +191,7 @@ public class NSGAIISystemTest extends SystemTestBase
 		c9.setFitness(fitnessFunctions.get(0), 0.6);
 		c10.setFitness(fitnessFunctions.get(0), 0.8);
 
-		List<NSGAChromosome> population = new ArrayList<NSGAChromosome>();
+		List<NSGAChromosome> population = new ArrayList<>();
 		population.add(c1);
 		population.add(c2);
 		population.add(c3);
@@ -203,12 +203,12 @@ public class NSGAIISystemTest extends SystemTestBase
 		population.add(c9);
 		population.add(c10);
 
-		CrowdingDistance<NSGAChromosome> crowdingDistance = new CrowdingDistance<NSGAChromosome>();
+		CrowdingDistance<NSGAChromosome> crowdingDistance = new CrowdingDistance<>();
 		crowdingDistance.crowdingDistanceAssignment(population, fitnessFunctions);
-		Collections.sort(population, new RankAndCrowdingDistanceComparator(true));
+		population.sort(new RankAndCrowdingDistanceComparator<>(true));
 
-		Assert.assertTrue(population.get(0).getDistance() == Double.POSITIVE_INFINITY);
-		Assert.assertTrue(population.get(1).getDistance() == Double.POSITIVE_INFINITY);
+		Assert.assertEquals(population.get(0).getDistance(), Double.POSITIVE_INFINITY, 0.0);
+		Assert.assertEquals(population.get(1).getDistance(), Double.POSITIVE_INFINITY, 0.0);
 
 		double epsilon = 1e-10;
 		Assert.assertTrue(Math.abs(0.25 - population.get(2).getDistance()) < epsilon);
@@ -224,9 +224,9 @@ public class NSGAIISystemTest extends SystemTestBase
 	@Test
     public void testCrowingDistanceAssignment_SeveralVariables()
 	{
-        NSGAII<NSGAChromosome> ga = new NSGAII<NSGAChromosome>(null);
+        NSGAII<NSGAChromosome> ga = new NSGAII<>(null);
 
-        Problem p = new SCH();
+        Problem<NSGAChromosome> p = new SCH();
         List<FitnessFunction<NSGAChromosome>> fitnessFunctions = p.getFitnessFunctions();
         ga.addFitnessFunctions(fitnessFunctions);
 
@@ -265,7 +265,7 @@ public class NSGAIISystemTest extends SystemTestBase
         c9.setFitness(fitnessFunctions.get(1), 0.7);
         c10.setFitness(fitnessFunctions.get(1), 0.9);
 
-        List<NSGAChromosome> population = new ArrayList<NSGAChromosome>();
+        List<NSGAChromosome> population = new ArrayList<>();
         population.add(c1);
         population.add(c2);
         population.add(c3);
@@ -277,12 +277,12 @@ public class NSGAIISystemTest extends SystemTestBase
         population.add(c9);
         population.add(c10);
 
-        CrowdingDistance<NSGAChromosome> crowdingDistance = new CrowdingDistance<NSGAChromosome>();
+        CrowdingDistance<NSGAChromosome> crowdingDistance = new CrowdingDistance<>();
         crowdingDistance.crowdingDistanceAssignment(population, fitnessFunctions);
-        Collections.sort(population, new RankAndCrowdingDistanceComparator(true));
+        population.sort(new RankAndCrowdingDistanceComparator<>(true));
 
-        Assert.assertTrue(population.get(0).getDistance() == Double.POSITIVE_INFINITY);
-        Assert.assertTrue(population.get(1).getDistance() == Double.POSITIVE_INFINITY);
+		Assert.assertEquals(population.get(0).getDistance(), Double.POSITIVE_INFINITY, 0.0);
+		Assert.assertEquals(population.get(1).getDistance(), Double.POSITIVE_INFINITY, 0.0);
 
         double epsilon = 1e-10;
         Assert.assertTrue(Math.abs(0.5 - population.get(2).getDistance()) < epsilon);
@@ -321,13 +321,14 @@ public class NSGAIISystemTest extends SystemTestBase
         Object result = evosuite.parseCommandLine(command);
         Assert.assertNotNull(result);
 
-        GeneticAlgorithm<?> ga = getGAFromResult(result);
+        GeneticAlgorithm<NSGAChromosome> ga =
+				(GeneticAlgorithm<NSGAChromosome>) getGAFromResult(result);
 
-        final FitnessFunction rho = ga.getFitnessFunctions().get(0);
-        final FitnessFunction ag = ga.getFitnessFunctions().get(1);
+        final FitnessFunction<NSGAChromosome> rho = ga.getFitnessFunctions().get(0);
+        final FitnessFunction<NSGAChromosome> ag = ga.getFitnessFunctions().get(1);
 
-        List<Chromosome> population = new ArrayList<Chromosome>(ga.getBestIndividuals());
-        Collections.sort(population, new SortByFitness(rho, false));
+        List<NSGAChromosome> population = new ArrayList<>(ga.getBestIndividuals());
+        population.sort(new SortByFitness<>(rho, false));
 
         Assert.assertEquals(0.0, population.get(0).getFitness(rho), 0.0);
         Assert.assertEquals(0.0, population.get(0).getFitness(ag), 0.0);

--- a/master/src/test/java/org/evosuite/ga/problems/multiobjective/POLIntTest.java
+++ b/master/src/test/java/org/evosuite/ga/problems/multiobjective/POLIntTest.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright (C) 2010-2018 Gordon Fraser, Andrea Arcuri and EvoSuite
  * contributors
  *
@@ -42,6 +42,8 @@ import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 
+import static java.util.Comparator.comparingDouble;
+
 /**
  * 
  * @author José Campos
@@ -53,15 +55,15 @@ public class POLIntTest
         Properties.POPULATION = 100;
         Properties.SEARCH_BUDGET = 10000;
         Properties.CROSSOVER_RATE = 0.9;
-        Properties.RANDOM_SEED = 1l;
+        Properties.RANDOM_SEED = 1L;
     }
 
     @Test
     public void testPOLFitnesses()
     {
-        Problem p = new POL();
-        FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
-        FitnessFunction f2 = (FitnessFunction) p.getFitnessFunctions().get(1);
+        Problem<NSGAChromosome> p = new POL();
+        FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
+        FitnessFunction<NSGAChromosome> f2 = p.getFitnessFunctions().get(1);
 
         double[] values = {-2.9272124303, 2.7365080818};
         NSGAChromosome c = new NSGAChromosome(-Math.PI, Math.PI, values);
@@ -81,37 +83,33 @@ public class POLIntTest
     {
         Properties.MUTATION_RATE = 1d / 2d;
 
-        ChromosomeFactory<?> factory = new RandomFactory(false, 2, -Math.PI, Math.PI);
+        ChromosomeFactory<NSGAChromosome> factory = new RandomFactory(false, 2, -Math.PI, Math.PI);
 
-        GeneticAlgorithm<?> ga = new NSGAII(factory);
-        BinaryTournamentSelectionCrowdedComparison ts = new BinaryTournamentSelectionCrowdedComparison();
+        GeneticAlgorithm<NSGAChromosome> ga = new NSGAII<>(factory);
+        BinaryTournamentSelectionCrowdedComparison<NSGAChromosome> ts =
+                new BinaryTournamentSelectionCrowdedComparison<>();
         ga.setSelectionFunction(ts);
         ga.setCrossOverFunction(new SBXCrossover());
 
-        Problem p = new POL();
-        final FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
-        final FitnessFunction f2 = (FitnessFunction) p.getFitnessFunctions().get(1);
+        Problem<NSGAChromosome> p = new POL();
+        final FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
+        final FitnessFunction<NSGAChromosome> f2 = p.getFitnessFunctions().get(1);
         ga.addFitnessFunction(f1);
         ga.addFitnessFunction(f2);
 
         // execute
         ga.generateSolution();
 
-        List<Chromosome> chromosomes = (List<Chromosome>) ga.getPopulation();
-        Collections.sort(chromosomes, new Comparator<Chromosome>() {
-            @Override
-            public int compare(Chromosome arg0, Chromosome arg1) {
-                return Double.compare(arg0.getFitness(f1), arg1.getFitness(f1));
-            }
-        });
+        List<NSGAChromosome> chromosomes = ga.getPopulation();
+        chromosomes.sort(comparingDouble(chr -> chr.getFitness(f1)));
 
         double[][] front = new double[Properties.POPULATION][2];
         int index = 0;
 
-        for (Chromosome chromosome : chromosomes) {
+        for (NSGAChromosome chromosome : chromosomes) {
             System.out.printf("%f,%f\n", chromosome.getFitness(f1), chromosome.getFitness(f2));
-            front[index][0] = Double.valueOf(chromosome.getFitness(f1));
-            front[index][1] = Double.valueOf(chromosome.getFitness(f2));
+            front[index][0] = chromosome.getFitness(f1);
+            front[index][1] = chromosome.getFitness(f2);
 
             index++;
         }

--- a/master/src/test/java/org/evosuite/ga/problems/multiobjective/SCH2.java
+++ b/master/src/test/java/org/evosuite/ga/problems/multiobjective/SCH2.java
@@ -34,9 +34,9 @@ import org.evosuite.ga.variables.DoubleVariable;
  * @author José Campos
  */
 @SuppressWarnings({ "serial" })
-public class SCH2<T extends NSGAChromosome> implements Problem
+public class SCH2 implements Problem<NSGAChromosome>
 {
-	private List<FitnessFunction<T>> fitnessFunctions = new ArrayList<FitnessFunction<T>>();
+	private final List<FitnessFunction<NSGAChromosome>> fitnessFunctions = new ArrayList<>();
 
 	public SCH2() {
 		super();
@@ -44,12 +44,11 @@ public class SCH2<T extends NSGAChromosome> implements Problem
 		/**
 		 * First fitness function
 		 */
-		class f1FitnessFunction extends FitnessFunction {
+		class f1FitnessFunction extends FitnessFunction<NSGAChromosome> {
 			@Override
-			public double getFitness(Chromosome c) {
-				NSGAChromosome individual = (NSGAChromosome)c;
+			public double getFitness(NSGAChromosome c) {
 
-				double x = ((DoubleVariable) individual.getVariables().get(0)).getValue();
+				double x = ((DoubleVariable) c.getVariables().get(0)).getValue();
 
 				double fitness = 0.0;
 				if (x <= 1.0)
@@ -61,7 +60,7 @@ public class SCH2<T extends NSGAChromosome> implements Problem
 				else if (x > 4.0)
                     fitness = x - 4.0;
 
-				updateIndividual(individual, fitness);
+				updateIndividual(c, fitness);
 				return fitness;
 			}
 			@Override
@@ -73,15 +72,14 @@ public class SCH2<T extends NSGAChromosome> implements Problem
 		/**
 		 * Second fitness function
 		 */
-		class f2FitnessFunction extends FitnessFunction {
+		class f2FitnessFunction extends FitnessFunction<NSGAChromosome> {
 			@Override
-			public double getFitness(Chromosome c) {
-				NSGAChromosome individual = (NSGAChromosome)c;
+			public double getFitness(NSGAChromosome c) {
 
-				double x = ((DoubleVariable) individual.getVariables().get(0)).getValue();
+				double x = ((DoubleVariable) c.getVariables().get(0)).getValue();
 				double fitness = Math.pow(x - 5.0, 2.0);
 
-				updateIndividual(individual, fitness);
+				updateIndividual(c, fitness);
 				return fitness;
 			}
 			@Override
@@ -95,7 +93,7 @@ public class SCH2<T extends NSGAChromosome> implements Problem
 	}
 
 	@Override
-	public List<FitnessFunction<T>> getFitnessFunctions() {
+	public List<FitnessFunction<NSGAChromosome>> getFitnessFunctions() {
 		return this.fitnessFunctions;
 	}
 }

--- a/master/src/test/java/org/evosuite/ga/problems/multiobjective/SCH2IntTest.java
+++ b/master/src/test/java/org/evosuite/ga/problems/multiobjective/SCH2IntTest.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright (C) 2010-2018 Gordon Fraser, Andrea Arcuri and EvoSuite
  * contributors
  *
@@ -43,6 +43,8 @@ import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 
+import static java.util.Comparator.comparingDouble;
+
 /**
  * 
  * @author José Campos
@@ -54,15 +56,15 @@ public class SCH2IntTest
         Properties.POPULATION = 100;
         Properties.SEARCH_BUDGET = 10000;
         Properties.CROSSOVER_RATE = 0.9;
-        Properties.RANDOM_SEED = 1l;
+        Properties.RANDOM_SEED = 1L;
     }
 
     @Test
     public void testSCH2Fitnesses()
     {
-        Problem p = new SCH2();
-        FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
-        FitnessFunction f2 = (FitnessFunction) p.getFitnessFunctions().get(1);
+        Problem<NSGAChromosome> p = new SCH2();
+        FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
+        FitnessFunction<NSGAChromosome> f2 = p.getFitnessFunctions().get(1);
 
         double[] values_n = {-3.0};
         NSGAChromosome c = new NSGAChromosome(-5.0, 10.0, values_n);
@@ -94,38 +96,34 @@ public class SCH2IntTest
     {
         Properties.MUTATION_RATE = 1d / 1d;
 
-        ChromosomeFactory<?> factory = new RandomFactory(false, 1, -5.0, 10.0);
+        ChromosomeFactory<NSGAChromosome> factory = new RandomFactory(false, 1, -5.0, 10.0);
 
-        GeneticAlgorithm<?> ga = new NSGAII(factory);
-        BinaryTournamentSelectionCrowdedComparison ts = new BinaryTournamentSelectionCrowdedComparison();
+        GeneticAlgorithm<NSGAChromosome> ga = new NSGAII<>(factory);
+        BinaryTournamentSelectionCrowdedComparison<NSGAChromosome> ts =
+                new BinaryTournamentSelectionCrowdedComparison<>();
         ts.setMaximize(false);
         ga.setSelectionFunction(ts);
         ga.setCrossOverFunction(new SBXCrossover());
 
-        Problem p = new SCH2();
-        final FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
-        final FitnessFunction f2 = (FitnessFunction) p.getFitnessFunctions().get(1);
+        Problem<NSGAChromosome> p = new SCH2();
+        final FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
+        final FitnessFunction<NSGAChromosome> f2 = p.getFitnessFunctions().get(1);
         ga.addFitnessFunction(f1);
         ga.addFitnessFunction(f2);
 
         // execute
         ga.generateSolution();
 
-        List<Chromosome> chromosomes = (List<Chromosome>) ga.getPopulation();
-        Collections.sort(chromosomes, new Comparator<Chromosome>() {
-            @Override
-            public int compare(Chromosome arg0, Chromosome arg1) {
-                return Double.compare(arg0.getFitness(f1), arg1.getFitness(f1));
-            }
-        });
+        List<NSGAChromosome> chromosomes = ga.getPopulation();
+        chromosomes.sort(comparingDouble(chr -> chr.getFitness(f1)));
 
         double[][] front = new double[Properties.POPULATION][2];
         int index = 0;
 
-        for (Chromosome chromosome : chromosomes) {
+        for (NSGAChromosome chromosome : chromosomes) {
             System.out.printf("%f,%f\n", chromosome.getFitness(f1), chromosome.getFitness(f2));
-            front[index][0] = Double.valueOf(chromosome.getFitness(f1));
-            front[index][1] = Double.valueOf(chromosome.getFitness(f2));
+            front[index][0] = chromosome.getFitness(f1);
+            front[index][1] = chromosome.getFitness(f2);
 
             index++;
         }

--- a/master/src/test/java/org/evosuite/ga/problems/multiobjective/ZDT4IntTest.java
+++ b/master/src/test/java/org/evosuite/ga/problems/multiobjective/ZDT4IntTest.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright (C) 2010-2018 Gordon Fraser, Andrea Arcuri and EvoSuite
  * contributors
  *
@@ -43,6 +43,8 @@ import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 
+import static java.util.Comparator.comparingDouble;
+
 /**
  * 
  * @author José Campos
@@ -54,15 +56,15 @@ public class ZDT4IntTest
         Properties.POPULATION = 100;
         Properties.SEARCH_BUDGET = 10000;
         Properties.CROSSOVER_RATE = 0.9;
-        Properties.RANDOM_SEED = 1l;
+        Properties.RANDOM_SEED = 1L;
     }
 
     @Test
     public void testZDT4Fitnesses()
     {
-        Problem p = new ZDT4();
-        FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
-        FitnessFunction f2 = (FitnessFunction) p.getFitnessFunctions().get(1);
+        Problem<NSGAChromosome> p = new ZDT4();
+        FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
+        FitnessFunction<NSGAChromosome> f2 = p.getFitnessFunctions().get(1);
 
         double[] values = {0.5, -5.0, -4.0, -3.0, -2.0, -1.0, 0.0, 1.0, 2.0, 3.0};
         NSGAChromosome c = new NSGAChromosome(-5, 5, values);
@@ -92,38 +94,33 @@ public class ZDT4IntTest
     {
         Properties.MUTATION_RATE = 1d / 10d;
 
-        ChromosomeFactory<?> factory = new RandomFactory(true, 10, -5.0, 5.0);
-
-        GeneticAlgorithm<?> ga = new NSGAII(factory);
-        BinaryTournamentSelectionCrowdedComparison ts = new BinaryTournamentSelectionCrowdedComparison();
+        ChromosomeFactory<NSGAChromosome> factory = new RandomFactory(true, 10, -5.0, 5.0);
+        GeneticAlgorithm<NSGAChromosome> ga = new NSGAII<>(factory);
+        BinaryTournamentSelectionCrowdedComparison<NSGAChromosome> ts =
+                new BinaryTournamentSelectionCrowdedComparison<>();
         ts.setMaximize(false);
         ga.setSelectionFunction(ts);
         ga.setCrossOverFunction(new SBXCrossover());
 
-        Problem p = new ZDT4();
-        final FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
-        final FitnessFunction f2 = (FitnessFunction) p.getFitnessFunctions().get(1);
+        Problem<NSGAChromosome> p = new ZDT4();
+        final FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
+        final FitnessFunction<NSGAChromosome> f2 = p.getFitnessFunctions().get(1);
         ga.addFitnessFunction(f1);
         ga.addFitnessFunction(f2);
 
         // execute
         ga.generateSolution();
 
-        List<Chromosome> chromosomes = (List<Chromosome>) ga.getPopulation();
-        Collections.sort(chromosomes, new Comparator<Chromosome>() {
-            @Override
-            public int compare(Chromosome arg0, Chromosome arg1) {
-                return Double.compare(arg0.getFitness(f1), arg1.getFitness(f1));
-            }
-        });
+        List<NSGAChromosome> chromosomes = ga.getPopulation();
+        chromosomes.sort(comparingDouble(arg0 -> arg0.getFitness(f1)));
 
         double[][] front = new double[Properties.POPULATION][2];
         int index = 0;
 
-        for (Chromosome chromosome : chromosomes) {
+        for (NSGAChromosome chromosome : chromosomes) {
             System.out.printf("%f,%f\n", chromosome.getFitness(f1), chromosome.getFitness(f2));
-            front[index][0] = Double.valueOf(chromosome.getFitness(f1));
-            front[index][1] = Double.valueOf(chromosome.getFitness(f2));
+            front[index][0] = chromosome.getFitness(f1);
+            front[index][1] = chromosome.getFitness(f2);
 
             index++;
         }

--- a/master/src/test/java/org/evosuite/ga/problems/multiobjective/ZDT6IntTest.java
+++ b/master/src/test/java/org/evosuite/ga/problems/multiobjective/ZDT6IntTest.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright (C) 2010-2018 Gordon Fraser, Andrea Arcuri and EvoSuite
  * contributors
  *
@@ -42,6 +42,8 @@ import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 
+import static java.util.Comparator.*;
+
 /**
  * 
  * @author José Campos
@@ -53,15 +55,15 @@ public class ZDT6IntTest
         Properties.POPULATION = 100;
         Properties.SEARCH_BUDGET = 50000;
         Properties.CROSSOVER_RATE = 0.9;
-        Properties.RANDOM_SEED = 1l;
+        Properties.RANDOM_SEED = 1L;
     }
 
     @Test
     public void testZDT6Fitnesses()
     {
-        Problem p = new ZDT6();
-        FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
-        FitnessFunction f2 = (FitnessFunction) p.getFitnessFunctions().get(1);
+        Problem<NSGAChromosome> p = new ZDT6();
+        FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
+        FitnessFunction<NSGAChromosome> f2 = p.getFitnessFunctions().get(1);
 
         double[] values = {0.541, 0.585, 0.915, 0.624, 0.493, 0.142, 0.971, 0.836, 0.763, 0.323};
         NSGAChromosome c = new NSGAChromosome(0.0, 1.0, values);
@@ -81,38 +83,34 @@ public class ZDT6IntTest
     {
         Properties.MUTATION_RATE = 1d / 10d;
 
-        ChromosomeFactory<?> factory = new RandomFactory(false, 10, 0.0, 1.0);
+        ChromosomeFactory<NSGAChromosome> factory = new RandomFactory(false, 10, 0.0, 1.0);
 
-        GeneticAlgorithm<?> ga = new NSGAII(factory);
-        BinaryTournamentSelectionCrowdedComparison ts = new BinaryTournamentSelectionCrowdedComparison();
+        GeneticAlgorithm<NSGAChromosome> ga = new NSGAII<>(factory);
+        BinaryTournamentSelectionCrowdedComparison<NSGAChromosome> ts =
+                new BinaryTournamentSelectionCrowdedComparison<>();
         ts.setMaximize(false);
         ga.setSelectionFunction(ts);
         ga.setCrossOverFunction(new SBXCrossover());
 
-        Problem p = new ZDT6();
-        final FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
-        final FitnessFunction f2 = (FitnessFunction) p.getFitnessFunctions().get(1);
+        Problem<NSGAChromosome> p = new ZDT6();
+        final FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
+        final FitnessFunction<NSGAChromosome> f2 = p.getFitnessFunctions().get(1);
         ga.addFitnessFunction(f1);
         ga.addFitnessFunction(f2);
 
         // execute
         ga.generateSolution();
 
-        List<Chromosome> chromosomes = (List<Chromosome>) ga.getPopulation();
-        Collections.sort(chromosomes, new Comparator<Chromosome>() {
-            @Override
-            public int compare(Chromosome arg0, Chromosome arg1) {
-                return Double.compare(arg0.getFitness(f1), arg1.getFitness(f1));
-            }
-        });
+        List<NSGAChromosome> chromosomes = ga.getPopulation();
+        chromosomes.sort(comparingDouble(chr -> chr.getFitness(f1)));
 
         double[][] front = new double[Properties.POPULATION][2];
         int index = 0;
 
-        for (Chromosome chromosome : chromosomes) {
+        for (NSGAChromosome chromosome : chromosomes) {
             System.out.printf("%f,%f\n", chromosome.getFitness(f1), chromosome.getFitness(f2));
-            front[index][0] = Double.valueOf(chromosome.getFitness(f1));
-            front[index][1] = Double.valueOf(chromosome.getFitness(f2));
+            front[index][0] = chromosome.getFitness(f1);
+            front[index][1] = chromosome.getFitness(f2);
 
             index++;
         }

--- a/master/src/test/java/org/evosuite/ga/problems/singleobjective/Beales.java
+++ b/master/src/test/java/org/evosuite/ga/problems/singleobjective/Beales.java
@@ -34,9 +34,9 @@ import org.evosuite.ga.variables.DoubleVariable;
  * @author José Campos
  */
 @SuppressWarnings({ "serial" })
-public class Beales<T extends NSGAChromosome> implements Problem
+public class Beales implements Problem<NSGAChromosome>
 {
-    private List<FitnessFunction<T>> fitnessFunctions = new ArrayList<FitnessFunction<T>>();
+    private final List<FitnessFunction<NSGAChromosome>> fitnessFunctions = new ArrayList<>();
 
     public Beales()
     {
@@ -45,18 +45,17 @@ public class Beales<T extends NSGAChromosome> implements Problem
         /**
          * Fitness function
          */
-        class fFitnessFunction extends FitnessFunction {
+        class fFitnessFunction extends FitnessFunction<NSGAChromosome> {
             @Override
-            public double getFitness(Chromosome c) {
-                NSGAChromosome individual = (NSGAChromosome)c;
+            public double getFitness(NSGAChromosome c) {
 
-                double x = ((DoubleVariable) individual.getVariables().get(0)).getValue();
-                double y = ((DoubleVariable) individual.getVariables().get(1)).getValue();
+                double x = ((DoubleVariable) c.getVariables().get(0)).getValue();
+                double y = ((DoubleVariable) c.getVariables().get(1)).getValue();
 
                 double fitness = Math.pow(1.5 - x + x * y, 2.0) +
                                  Math.pow(2.25 - x + Math.pow(x * y, 2.0), 2.0) +
                                  Math.pow(2.625 - x + Math.pow(x * y, 3.0), 2.0);
-                updateIndividual(individual, fitness);
+                updateIndividual(c, fitness);
                 return fitness;
             }
             @Override
@@ -69,7 +68,7 @@ public class Beales<T extends NSGAChromosome> implements Problem
     }
 
     @Override
-    public List getFitnessFunctions() {
+    public List<FitnessFunction<NSGAChromosome>> getFitnessFunctions() {
         return this.fitnessFunctions;
     }
 }

--- a/master/src/test/java/org/evosuite/ga/problems/singleobjective/Booths.java
+++ b/master/src/test/java/org/evosuite/ga/problems/singleobjective/Booths.java
@@ -34,9 +34,9 @@ import org.evosuite.ga.variables.DoubleVariable;
  * @author José Campos
  */
 @SuppressWarnings({ "serial" })
-public class Booths<T extends NSGAChromosome> implements Problem
+public class Booths implements Problem<NSGAChromosome>
 {
-    private List<FitnessFunction<T>> fitnessFunctions = new ArrayList<FitnessFunction<T>>();
+    private final List<FitnessFunction<NSGAChromosome>> fitnessFunctions = new ArrayList<>();
 
     public Booths()
     {
@@ -45,16 +45,14 @@ public class Booths<T extends NSGAChromosome> implements Problem
         /**
          * Fitness function
          */
-        class fFitnessFunction extends FitnessFunction {
+        class fFitnessFunction extends FitnessFunction<NSGAChromosome> {
             @Override
-            public double getFitness(Chromosome c) {
-                NSGAChromosome individual = (NSGAChromosome)c;
-
-                double x = ((DoubleVariable) individual.getVariables().get(0)).getValue();
-                double y = ((DoubleVariable) individual.getVariables().get(1)).getValue();
+            public double getFitness(NSGAChromosome c) {
+                double x = ((DoubleVariable) c.getVariables().get(0)).getValue();
+                double y = ((DoubleVariable) c.getVariables().get(1)).getValue();
 
                 double fitness = Math.pow(x + 2.0 * y - 7.0, 2.0) + Math.pow(2.0 * x + y - 5.0, 2.0);
-                updateIndividual(individual, fitness);
+                updateIndividual(c, fitness);
                 return fitness;
             }
             @Override
@@ -67,7 +65,7 @@ public class Booths<T extends NSGAChromosome> implements Problem
     }
 
     @Override
-    public List getFitnessFunctions() {
+    public List<FitnessFunction<NSGAChromosome>> getFitnessFunctions() {
         return this.fitnessFunctions;
     }
 }

--- a/master/src/test/java/org/evosuite/ga/problems/singleobjective/TestBeales.java
+++ b/master/src/test/java/org/evosuite/ga/problems/singleobjective/TestBeales.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright (C) 2010-2018 Gordon Fraser, Andrea Arcuri and EvoSuite
  * contributors
  *
@@ -40,6 +40,8 @@ import org.junit.Assert;
 import org.junit.BeforeClass;
 import org.junit.Test;
 
+import static java.util.Comparator.comparingDouble;
+
 public class TestBeales
 {
     @BeforeClass
@@ -47,14 +49,14 @@ public class TestBeales
         Properties.POPULATION = 100;
         Properties.SEARCH_BUDGET = 250;
         Properties.CROSSOVER_RATE = 0.9;
-        Properties.RANDOM_SEED = 1l;
+        Properties.RANDOM_SEED = 1L;
     }
 
     @Test
     public void testBealesFitness()
     {
-        Problem p = new Beales();
-        FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
+        Problem<NSGAChromosome> p = new Beales();
+        FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
 
         double[] values = {-2.0, 1.0};
         NSGAChromosome c = new NSGAChromosome(-4.5, 4.5, values);
@@ -75,38 +77,31 @@ public class TestBeales
     {
         Properties.MUTATION_RATE = 1d / 2d;
 
-        ChromosomeFactory<?> factory = new RandomFactory(false, 2, -4.5, 4.5);
+        ChromosomeFactory<NSGAChromosome> factory = new RandomFactory(false, 2, -4.5, 4.5);
 
-        //GeneticAlgorithm<?> ga = new NSGAII(factory);
-        GeneticAlgorithm<?> ga = new NSGAII(factory);
-        BinaryTournamentSelectionCrowdedComparison ts = new BinaryTournamentSelectionCrowdedComparison();
+        GeneticAlgorithm<NSGAChromosome> ga = new NSGAII<>(factory);
+        BinaryTournamentSelectionCrowdedComparison<NSGAChromosome> ts =
+                new BinaryTournamentSelectionCrowdedComparison<>();
         //BinaryTournament ts = new BinaryTournament();
         ga.setSelectionFunction(ts);
         ga.setCrossOverFunction(new SBXCrossover());
 
-        Problem p = new Beales();
-        final FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
+        Problem<NSGAChromosome> p = new Beales();
+        final FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
         ga.addFitnessFunction(f1);
 
         // execute
         ga.generateSolution();
 
-        List<Chromosome> chromosomes = (List<Chromosome>) ga.getPopulation();
-        Collections.sort(chromosomes, new Comparator<Chromosome>() {
-            @Override
-            public int compare(Chromosome arg0, Chromosome arg1) {
-                return Double.compare(arg0.getFitness(f1), arg1.getFitness(f1));
-            }
-        });
+        List<NSGAChromosome> chromosomes = ga.getPopulation();
+        chromosomes.sort(comparingDouble(chr -> chr.getFitness(f1)));
 
-        for (Chromosome chromosome : chromosomes)
+        for (NSGAChromosome chromosome : chromosomes)
             Assert.assertEquals(chromosome.getFitness(f1), 0.29, 0.01);
 
-        for (Chromosome chromosome : chromosomes) {
-            NSGAChromosome nsga_c = (NSGAChromosome)chromosome;
-
-            DoubleVariable x = (DoubleVariable) nsga_c.getVariables().get(0);
-            DoubleVariable y = (DoubleVariable) nsga_c.getVariables().get(1);
+        for (NSGAChromosome chromosome : chromosomes) {
+            DoubleVariable x = (DoubleVariable) chromosome.getVariables().get(0);
+            DoubleVariable y = (DoubleVariable) chromosome.getVariables().get(1);
             System.out.printf("%f,%f : %f\n", x.getValue(), y.getValue(), chromosome.getFitness(f1));
         }
     }

--- a/master/src/test/java/org/evosuite/ga/problems/singleobjective/TestBooths.java
+++ b/master/src/test/java/org/evosuite/ga/problems/singleobjective/TestBooths.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright (C) 2010-2018 Gordon Fraser, Andrea Arcuri and EvoSuite
  * contributors
  *
@@ -40,6 +40,8 @@ import org.junit.Assert;
 import org.junit.BeforeClass;
 import org.junit.Test;
 
+import static java.util.Comparator.comparingDouble;
+
 public class TestBooths
 {
     @BeforeClass
@@ -47,14 +49,14 @@ public class TestBooths
         Properties.POPULATION = 100;
         Properties.SEARCH_BUDGET = 250;
         Properties.CROSSOVER_RATE = 0.9;
-        Properties.RANDOM_SEED = 1l;
+        Properties.RANDOM_SEED = 1L;
     }
 
     @Test
     public void testBoothsFitness()
     {
-        Problem p = new Booths();
-        FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
+        Problem<NSGAChromosome> p = new Booths();
+        FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
 
         double[] values = {-2.0, 1.0};
         NSGAChromosome c = new NSGAChromosome(-10.0, 10.0, values);
@@ -79,38 +81,31 @@ public class TestBooths
     {
         Properties.MUTATION_RATE = 1d / 2d;
 
-        ChromosomeFactory<?> factory = new RandomFactory(false, 2, -10.0, 10.0);
+        ChromosomeFactory<NSGAChromosome> factory = new RandomFactory(false, 2, -10.0, 10.0);
 
-        //GeneticAlgorithm<?> ga = new NSGAII(factory);
-        GeneticAlgorithm<?> ga = new NSGAII(factory);
-        BinaryTournamentSelectionCrowdedComparison ts = new BinaryTournamentSelectionCrowdedComparison(false);
+        GeneticAlgorithm<NSGAChromosome> ga = new NSGAII<>(factory);
+        BinaryTournamentSelectionCrowdedComparison<NSGAChromosome> ts =
+                new BinaryTournamentSelectionCrowdedComparison<>(false);
         //BinaryTournament ts = new BinaryTournament();
         ga.setSelectionFunction(ts);
         ga.setCrossOverFunction(new SBXCrossover());
 
-        Problem p = new Booths();
-        final FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
+        Problem<NSGAChromosome> p = new Booths();
+        final FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
         ga.addFitnessFunction(f1);
 
         // execute
         ga.generateSolution();
 
-        List<Chromosome> chromosomes = (List<Chromosome>) ga.getPopulation();
-        Collections.sort(chromosomes, new Comparator<Chromosome>() {
-            @Override
-            public int compare(Chromosome arg0, Chromosome arg1) {
-                return Double.compare(arg0.getFitness(f1), arg1.getFitness(f1));
-            }
-        });
+        List<NSGAChromosome> chromosomes = ga.getPopulation();
+        chromosomes.sort(comparingDouble(chr -> chr.getFitness(f1)));
 
-        for (Chromosome chromosome : chromosomes)
+        for (NSGAChromosome chromosome : chromosomes)
             Assert.assertEquals(chromosome.getFitness(f1), 0.000, 0.001);
 
-        for (Chromosome chromosome : chromosomes) {
-            NSGAChromosome nsga_c = (NSGAChromosome)chromosome;
-
-            DoubleVariable x = (DoubleVariable) nsga_c.getVariables().get(0);
-            DoubleVariable y = (DoubleVariable) nsga_c.getVariables().get(1);
+        for (NSGAChromosome chromosome : chromosomes) {
+            DoubleVariable x = (DoubleVariable) chromosome.getVariables().get(0);
+            DoubleVariable y = (DoubleVariable) chromosome.getVariables().get(1);
             System.out.printf("%f,%f : %f\n", x.getValue(), y.getValue(), chromosome.getFitness(f1));
         }
     }

--- a/master/src/test/java/org/evosuite/ga/problems/singleobjective/TestShere.java
+++ b/master/src/test/java/org/evosuite/ga/problems/singleobjective/TestShere.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright (C) 2010-2018 Gordon Fraser, Andrea Arcuri and EvoSuite
  * contributors
  *
@@ -40,6 +40,8 @@ import org.junit.Assert;
 import org.junit.BeforeClass;
 import org.junit.Test;
 
+import static java.util.Comparator.comparingDouble;
+
 public class TestShere
 {
     @BeforeClass
@@ -47,14 +49,14 @@ public class TestShere
         Properties.POPULATION = 100;
         Properties.SEARCH_BUDGET = 2500;
         Properties.CROSSOVER_RATE = 0.9;
-        Properties.RANDOM_SEED = 1l;
+        Properties.RANDOM_SEED = 1L;
     }
 
     @Test
     public void testSphereFitness()
     {
-        Problem p = new Sphere();
-        FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
+        Problem<NSGAChromosome> p = new Sphere();
+        FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
 
         double[] values = {-2.0};
         NSGAChromosome c = new NSGAChromosome(Math.pow(-10.0, 3.0), Math.pow(10.0, 3.0), values);
@@ -74,37 +76,31 @@ public class TestShere
     {
         Properties.MUTATION_RATE = 1d / 1d;
 
-        ChromosomeFactory<?> factory = new RandomFactory(false, 1, Math.pow(-10.0, 3.0), Math.pow(10.0, 3.0));
+        ChromosomeFactory<NSGAChromosome> factory = new RandomFactory(false, 1, Math.pow(-10.0, 3.0),
+                Math.pow(10.0, 3.0));
 
-        //GeneticAlgorithm<?> ga = new NSGAII(factory);
-        GeneticAlgorithm<?> ga = new NSGAII(factory);
-        BinaryTournamentSelectionCrowdedComparison ts = new BinaryTournamentSelectionCrowdedComparison();
+        GeneticAlgorithm<NSGAChromosome> ga = new NSGAII<>(factory);
+        BinaryTournamentSelectionCrowdedComparison<NSGAChromosome> ts =
+                new BinaryTournamentSelectionCrowdedComparison<>();
         //BinaryTournament ts = new BinaryTournament();
         ga.setSelectionFunction(ts);
         ga.setCrossOverFunction(new SBXCrossover());
 
-        Problem p = new Sphere();
-        final FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
+        Problem<NSGAChromosome> p = new Sphere();
+        final FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
         ga.addFitnessFunction(f1);
 
         // execute
         ga.generateSolution();
 
-        List<Chromosome> chromosomes = (List<Chromosome>) ga.getPopulation();
-        Collections.sort(chromosomes, new Comparator<Chromosome>() {
-            @Override
-            public int compare(Chromosome arg0, Chromosome arg1) {
-                return Double.compare(arg0.getFitness(f1), arg1.getFitness(f1));
-            }
-        });
+        List<NSGAChromosome> chromosomes = ga.getPopulation();
+        chromosomes.sort(comparingDouble(chr -> chr.getFitness(f1)));
 
-        for (Chromosome chromosome : chromosomes)
+        for (NSGAChromosome chromosome : chromosomes)
             Assert.assertEquals(chromosome.getFitness(f1), 0.00, 0.01);
 
-        for (Chromosome chromosome : chromosomes) {
-            NSGAChromosome nsga_c = (NSGAChromosome)chromosome;
-
-            DoubleVariable x = (DoubleVariable) nsga_c.getVariables().get(0);
+        for (NSGAChromosome chromosome : chromosomes) {
+            DoubleVariable x = (DoubleVariable) chromosome.getVariables().get(0);
             System.out.printf("%f : %f\n", x.getValue(), chromosome.getFitness(f1));
         }
     }

--- a/master/src/test/java/org/evosuite/ga/problems/singleobjective/TestThreeHump.java
+++ b/master/src/test/java/org/evosuite/ga/problems/singleobjective/TestThreeHump.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright (C) 2010-2018 Gordon Fraser, Andrea Arcuri and EvoSuite
  * contributors
  *
@@ -40,6 +40,8 @@ import org.junit.Assert;
 import org.junit.BeforeClass;
 import org.junit.Test;
 
+import static java.util.Comparator.comparingDouble;
+
 public class TestThreeHump
 {
     @BeforeClass
@@ -47,14 +49,14 @@ public class TestThreeHump
         Properties.POPULATION = 100;
         Properties.SEARCH_BUDGET = 250;
         Properties.CROSSOVER_RATE = 0.9;
-        Properties.RANDOM_SEED = 1l;
+        Properties.RANDOM_SEED = 1L;
     }
 
     @Test
     public void testThreeHumpFitness()
     {
-        Problem p = new ThreeHump();
-        FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
+        Problem<NSGAChromosome> p = new ThreeHump();
+        FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
 
         double[] values = {-2.0, 3.0};
         NSGAChromosome c = new NSGAChromosome(-5.0, 5.0, values);
@@ -75,38 +77,31 @@ public class TestThreeHump
     {
         Properties.MUTATION_RATE = 1d / 2d;
 
-        ChromosomeFactory<?> factory = new RandomFactory(false, 2, -5.0, 5.0);
+        ChromosomeFactory<NSGAChromosome> factory = new RandomFactory(false, 2, -5.0, 5.0);
 
-        //GeneticAlgorithm<?> ga = new NSGAII(factory);
-        GeneticAlgorithm<?> ga = new NSGAII(factory);
-        BinaryTournamentSelectionCrowdedComparison ts = new BinaryTournamentSelectionCrowdedComparison();
+        GeneticAlgorithm<NSGAChromosome> ga = new NSGAII<>(factory);
+        BinaryTournamentSelectionCrowdedComparison<NSGAChromosome> ts =
+                new BinaryTournamentSelectionCrowdedComparison<>();
         //BinaryTournament ts = new BinaryTournament();
         ga.setSelectionFunction(ts);
         ga.setCrossOverFunction(new SBXCrossover());
 
-        Problem p = new ThreeHump();
-        final FitnessFunction f1 = (FitnessFunction) p.getFitnessFunctions().get(0);
+        Problem<NSGAChromosome> p = new ThreeHump();
+        final FitnessFunction<NSGAChromosome> f1 = p.getFitnessFunctions().get(0);
         ga.addFitnessFunction(f1);
 
         // execute
         ga.generateSolution();
 
-        List<Chromosome> chromosomes = (List<Chromosome>) ga.getPopulation();
-        Collections.sort(chromosomes, new Comparator<Chromosome>() {
-            @Override
-            public int compare(Chromosome arg0, Chromosome arg1) {
-                return Double.compare(arg0.getFitness(f1), arg1.getFitness(f1));
-            }
-        });
+        List<NSGAChromosome> chromosomes = ga.getPopulation();
+        chromosomes.sort(comparingDouble(chr -> chr.getFitness(f1)));
 
-        for (Chromosome chromosome : chromosomes)
+        for (NSGAChromosome chromosome : chromosomes)
             Assert.assertEquals(chromosome.getFitness(f1), 0.000, 0.001);
 
-        for (Chromosome chromosome : chromosomes) {
-            NSGAChromosome nsga_c = (NSGAChromosome)chromosome;
-
-            DoubleVariable x = (DoubleVariable) nsga_c.getVariables().get(0);
-            DoubleVariable y = (DoubleVariable) nsga_c.getVariables().get(1);
+        for (NSGAChromosome chromosome : chromosomes) {
+            DoubleVariable x = (DoubleVariable) chromosome.getVariables().get(0);
+            DoubleVariable y = (DoubleVariable) chromosome.getVariables().get(1);
             System.out.printf("%f,%f : %f\n", x.getValue(), y.getValue(), chromosome.getFitness(f1));
         }
     }

--- a/master/src/test/java/org/evosuite/ga/problems/singleobjective/ThreeHump.java
+++ b/master/src/test/java/org/evosuite/ga/problems/singleobjective/ThreeHump.java
@@ -34,9 +34,9 @@ import org.evosuite.ga.variables.DoubleVariable;
  * @author José Campos
  */
 @SuppressWarnings({ "serial" })
-public class ThreeHump<T extends NSGAChromosome> implements Problem
+public class ThreeHump implements Problem<NSGAChromosome>
 {
-    private List<FitnessFunction<T>> fitnessFunctions = new ArrayList<FitnessFunction<T>>();
+    private final List<FitnessFunction<NSGAChromosome>> fitnessFunctions = new ArrayList<>();
 
     public ThreeHump()
     {
@@ -45,20 +45,19 @@ public class ThreeHump<T extends NSGAChromosome> implements Problem
         /**
          * Fitness function
          */
-        class fFitnessFunction extends FitnessFunction {
+        class fFitnessFunction extends FitnessFunction<NSGAChromosome> {
             @Override
-            public double getFitness(Chromosome c) {
-                NSGAChromosome individual = (NSGAChromosome)c;
+            public double getFitness(NSGAChromosome c) {
 
-                double x = ((DoubleVariable) individual.getVariables().get(0)).getValue();
-                double y = ((DoubleVariable) individual.getVariables().get(1)).getValue();
+                double x = ((DoubleVariable) c.getVariables().get(0)).getValue();
+                double y = ((DoubleVariable) c.getVariables().get(1)).getValue();
 
                 double fitness = (2.0 * Math.pow(x, 2.0))
                                 - (1.05 * Math.pow(x, 4.0))
                                 + (Math.pow(x, 6.0) / 6.0)
                                 + (x * y)
                                 + (Math.pow(y, 2.0));
-                updateIndividual(individual, fitness);
+                updateIndividual(c, fitness);
                 return fitness;
             }
             @Override
@@ -71,7 +70,7 @@ public class ThreeHump<T extends NSGAChromosome> implements Problem
     }
 
     @Override
-    public List getFitnessFunctions() {
+    public List<FitnessFunction<NSGAChromosome>> getFitnessFunctions() {
         return this.fitnessFunctions;
     }
 }

